File: ASP .NET/Clients/Controllers/AuthController.cs
using Clients.Dtos;
using Clients.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Collections.Generic;

namespace Clients.Controllers
{
    /// <summary>
    /// Controlador optimizado de autenticación
    /// Maneja login, logout, verificación de email y operaciones de auth
    /// Equivalente a AuthController.java en la API JAVA
    /// </summary>
    [ApiController]
    [Route("api/[controller]")]
    public class AuthController : ControllerBase
    {
        private readonly IUserService _userService;
        private readonly IJwtProvider _jwtProvider;
        private readonly ILogger<AuthController> _logger;
        private readonly IConfiguration _configuration;
        private readonly ITokenBlacklistService _tokenBlacklist;

        public AuthController(
            IUserService userService, 
            IJwtProvider jwtProvider, 
            ILogger<AuthController> logger,
            IConfiguration configuration,
            ITokenBlacklistService tokenBlacklist)
        {
            _userService = userService;
            _jwtProvider = jwtProvider;
            _logger = logger;
            _configuration = configuration;
            _tokenBlacklist = tokenBlacklist;
            _logger = logger;
            _configuration = configuration;
        }

        /// <summary>
        /// POST /api/auth/login - Autentica al usuario
        /// Retorna token en el body y lo guarda en cookie (como en JAVA)
        /// </summary>
        [HttpPost("login")]
        public async Task<ActionResult<ApiResponse<UserDto>>> Login([FromBody] LoginRequest request)
        {
            _logger.LogInformation("📝 Intento de login para email: {Email}", request.Email);

            try
            {
                // Validar credenciales (lanza excepción si usuario no está verificado)
                bool isValid = await _userService.ValidateCredentialsAsync(request.Email, request.Password);

                if (!isValid)
                {
                    _logger.LogWarning("⚠️ Login fallido: credenciales inválidas para {Email}", request.Email);
                    return Unauthorized(ApiResponse<UserDto>.Unauthorized("Email o contraseña incorrectos"));
                }

                // Obtener usuario
                var user = await _userService.GetUserByEmailAsync(request.Email);
                if (user == null)
                {
                    return Unauthorized(ApiResponse<UserDto>.Unauthorized("Usuario no encontrado"));
                }

                // Generar token JWT
                string token = _jwtProvider.GenerateToken(user);

                // Agregar cookie de autenticación (como en JAVA)
                AddAuthCookie(token);

                _logger.LogInformation("✅ Login exitoso para usuario: {Email}", request.Email);

                // Retornar token en el body + datos del usuario como data (como en JAVA)
                var userDto = UserDto.FromEntity(user);
                var response = new ApiResponse<UserDto>
                {
                    Success = true,
                    Message = "Login exitoso",
                    Data = userDto,
                    StatusCode = 200,
                    Timestamp = DateTime.UtcNow
                };

                // Agregar el token al cuerpo de la respuesta (compatible con frontend)
                return Ok(new
                {
                    success = response.Success,
                    message = response.Message,
                    token = token,  // Frontend extrae el token de aquí
                    data = userDto,
                    statusCode = response.StatusCode,
                    timestamp = response.Timestamp
                });
            }
            catch (InvalidOperationException ex)
            {
                _logger.LogWarning("⚠️ Error de validación en login: {Message}", ex.Message);
                return Unauthorized(ApiResponse<UserDto>.Unauthorized(ex.Message));
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "❌ Error inesperado en login: {Message}", ex.Message);
                return StatusCode(500, ApiResponse<UserDto>.InternalServerError("Error en el proceso de login"));
            }
        }

        /// <summary>
        /// GET /api/auth/verify/{token} - Verifica el email del usuario
        /// Redirige a la URL configurada tras la verificación (como en JAVA)
        /// </summary>
        [HttpGet("verify/{token}")]
        public async Task<IActionResult> VerifyEmail(string token)
        {
            _logger.LogInformation("📧 Intento de verificación de email con token");

            try
            {
                await _userService.VerifyEmailAsync(token);

                // Redirigir a la URL de login del frontend configurada
                var frontendLoginUrl = _configuration["AppUrls:FrontendLoginUrl"] ?? "http://localhost/login";
                return Redirect($"{frontendLoginUrl}?verified=1");
            }
            catch (InvalidOperationException ex)
            {
                _logger.LogWarning("⚠️ Error en verificación de email: {Message}", ex.Message);
                var frontendRegisterUrl = _configuration["AppUrls:FrontendRegisterUrl"] ?? "http://localhost/register";
                return Redirect($"{frontendRegisterUrl}?verification=failed");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "❌ Error inesperado en verificación de email: {Message}", ex.Message);
                var frontendRegisterUrl = _configuration["AppUrls:FrontendRegisterUrl"] ?? "http://localhost/register";
                return Redirect($"{frontendRegisterUrl}?verification=failed");
            }
        }

        /// <summary>
        /// POST /api/auth/refresh - Refresca el JWT token
        /// </summary>
        [Authorize]
        [HttpPost("refresh")]
        public async Task<ActionResult<ApiResponse<object>>> RefreshToken()
        {
            _logger.LogInformation("🔄 Solicitud de refresco de token");

            try
            {
                // Obtener email del usuario autenticado
                var email = User.FindFirst(System.Security.Claims.ClaimTypes.NameIdentifier)?.Value
                    ?? User.FindFirst(System.Security.Claims.ClaimTypes.Email)?.Value;

                if (string.IsNullOrEmpty(email))
                {
                    return Unauthorized(ApiResponse<object>.Unauthorized("No autenticado"));
                }

                var user = await _userService.GetUserByEmailAsync(email);
                if (user == null)
                {
                    return Unauthorized(ApiResponse<object>.Unauthorized("Usuario no encontrado"));
                }

                // Generar nuevo token
                string newToken = _jwtProvider.GenerateToken(user);
                AddAuthCookie(newToken);

                _logger.LogInformation("✅ Token refrescado para usuario: {Email}", email);

                return Ok(new
                {
                    success = true,
                    message = "Token refrescado exitosamente",
                    data = new { token = newToken },
                    statusCode = 200,
                    timestamp = DateTime.UtcNow
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "❌ Error al refrescar token: {Message}", ex.Message);
                return StatusCode(500, ApiResponse<object>.InternalServerError("Error al refrescar token"));
            }
        }

        /// <summary>
        /// GET /api/auth/debug - Endpoint de debug para ver estado de cookies y tokens
        /// SOLO PARA DESARROLLO - Eliminar en producción
        /// </summary>
        [HttpGet("debug")]
        public IActionResult Debug()
        {
            _logger.LogInformation("🔍 DEBUG: Información de autenticación");

            // Obtener cookies
            var cookies = Request.Cookies.ToDictionary(c => c.Key, c => c.Value.Substring(0, Math.Min(50, c.Value.Length)) + "...");

            // Obtener headers
            var authHeader = Request.Headers["Authorization"].FirstOrDefault();
            var bearerToken = authHeader?.StartsWith("Bearer ") == true 
                ? authHeader.Substring("Bearer ".Length).Substring(0, Math.Min(50, authHeader.Length - 7)) + "..."
                : "None";

            // Obtener info de autenticación
            var isAuthenticated = User.Identity?.IsAuthenticated ?? false;
            var userEmail = User.FindFirst(System.Security.Claims.ClaimTypes.NameIdentifier)?.Value ?? "None";
            var userId = User.FindFirst("userId")?.Value ?? "None";

            // ✅ NUEVO: Verificar si el token actual está en blacklist
            string? fullToken = ExtractTokenFromRequest();
            bool isTokenBlacklisted = !string.IsNullOrEmpty(fullToken) && _tokenBlacklist.IsTokenBlacklisted(fullToken);

            return Ok(new
            {
                debug = true,
                timestamp = DateTime.UtcNow,
                cookies = cookies,
                authHeader = bearerToken,
                authentication = new
                {
                    isAuthenticated = isAuthenticated,
                    userEmail = userEmail,
                    userId = userId,
                    tokenBlacklisted = isTokenBlacklisted,
                    claims = User.Claims.Select(c => new { c.Type, c.Value }).ToList()
                },
                message = "Este es un endpoint de DEBUG. Úsalo para ver qué está pasando con cookies y tokens. Si tokenBlacklisted=true, el token fue invalidado en logout."
            });
        }

        [HttpPost("logout")]
        public IActionResult Logout()
        {
            _logger.LogInformation("🚪 Solicitud de logout recibida");

            // ✅ NUEVO: Extraer el token actual del request y añadirlo a la blacklist
            string? token = ExtractTokenFromRequest();
            if (!string.IsNullOrEmpty(token))
            {
                // Obtener la expiración del token para saber hasta cuándo invalidarlo
                DateTime tokenExpiration = _jwtProvider.GetTokenExpiration(token);
                _tokenBlacklist.BlacklistToken(token, tokenExpiration);
                _logger.LogInformation($"🚫 Token invalidado y añadido a blacklist hasta: {tokenExpiration:yyyy-MM-dd HH:mm:ss}");
            }
            else
            {
                _logger.LogWarning("⚠️ No se encontró token en el request de logout");
            }

            // Eliminar la cookie del navegador
            RemoveAuthCookie();

            // Agregar headers para indicar al frontend que limpie almacenamientos
            Response.Headers.Add("X-Clear-Auth", "true");
            Response.Headers.Add("Cache-Control", "no-store, no-cache, must-revalidate, proxy-revalidate");
            Response.Headers.Add("Pragma", "no-cache");
            Response.Headers.Add("Expires", "0");

            return Ok(new
            {
                success = true,
                message = "Logout exitoso",
                statusCode = 200,
                timestamp = DateTime.UtcNow,
                cleared = true
            });
        }

        /// <summary>
        /// Agrega cookie de autenticación segura (como en JAVA)
        /// </summary>
        private void AddAuthCookie(string token)
        {
            var cookieOptions = new CookieOptions
            {
                HttpOnly = true,
                Secure = false,  // En desarrollo. En producción: true
                SameSite = SameSiteMode.Lax,
                Path = "/",
                MaxAge = TimeSpan.FromSeconds(86400)  // 24 horas
            };

            Response.Cookies.Append("auth_token", token, cookieOptions);
            _logger.LogInformation("🍪 Cookie auth_token configurada (Secure=False, SameSite=Lax)");
        }

        /// <summary>
        /// Elimina la cookie de autenticación (como en JAVA)
        /// </summary>
        private void RemoveAuthCookie()
        {
            // IMPORTANTE: Eliminar con los MISMOS CookieOptions que se usaron al crear
            var cookieOptions = new CookieOptions
            {
                Path = "/",
                HttpOnly = true,
                Secure = false,  // Debe coincidir con AddAuthCookie
                SameSite = SameSiteMode.Lax,
                IsEssential = false
            };

            Response.Cookies.Delete("auth_token", cookieOptions);

            // Logging de la operación
            _logger.LogInformation("🚪 Cookie auth_token marcada para eliminación (Path=/,HttpOnly=True,Secure=False,SameSite=Lax)");
        }

        /// <summary>
        /// Extrae el token JWT del request actual (del header o de la cookie)
        /// </summary>
        private string? ExtractTokenFromRequest()
        {
            // Intentar obtener del header Authorization
            if (Request.Headers.TryGetValue("Authorization", out var authHeader))
            {
                var authHeaderValue = authHeader.ToString();
                if (authHeaderValue.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase))
                {
                    return authHeaderValue.Substring("Bearer ".Length).Trim();
                }
            }

            // Intentar obtener de la cookie auth_token
            if (Request.Cookies.TryGetValue("auth_token", out var token))
            {
                return token;
            }

            return null;
        }
    }
}

File: JAVA/Clients/clients/src/main/java/com/futureprograms/clients/controller/AuthController.java
package com.futureprograms.clients.controller;

import com.futureprograms.clients.config.JwtProvider;
import com.futureprograms.clients.dto.LoginRequest;
import com.futureprograms.clients.dto.UserDto;
import com.futureprograms.clients.entity.User;
import com.futureprograms.clients.service.UserService;
import com.futureprograms.clients.util.ApiConstants;
import com.futureprograms.clients.util.AuthenticationHelper;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Map;

/**
 * Controlador optimizado de autenticación
 * Maneja login, logout, verificación de email y operaciones de auth
 */
@RestController
@RequestMapping(ApiConstants.AUTH_ENDPOINT)
@Slf4j
@RequiredArgsConstructor
public class AuthController {

    private final UserService userService;
    private final JwtProvider jwtProvider;
    private final AuthenticationHelper authenticationHelper;

    @Value("${app.frontend.login-url:http://localhost/login}")
    private String frontendLoginUrl;

    @Value("${app.frontend.register-url:http://localhost/register}")
    private String frontendRegisterUrl;

    /**
     * POST /api/auth/login - Autentica al usuario
     */
    @PostMapping("/login")
    public ResponseEntity<?> login(
            @Valid @RequestBody LoginRequest request,
            HttpServletResponse response
    ) {
        if (!userService.validateCredentials(request.getEmail(), request.getPassword())) {
            throw new IllegalArgumentException(ApiConstants.ERR_INVALID_CREDENTIALS);
        }

        User user = userService.getUserByEmail(request.getEmail())
                .orElseThrow(() -> new IllegalStateException(ApiConstants.ERR_USER_NOT_FOUND));

        String token = jwtProvider.generateToken(user.getEmail());
        addAuthCookie(response, token);

        authenticationHelper.logAuthEvent(user.getEmail(), "login_successful");

        // Retornar token en el body + datos del usuario como data
        // Frontend extrae el token de response.token
        return ResponseEntity.ok()
                .body(Map.of(
                        "success", true,
                        "message", ApiConstants.MSG_LOGIN_SUCCESS,
                        "token", token,
                        "data", UserDto.fromEntity(user)
                ));
    }

    /**
     * GET /api/auth/verify/{token} - Verifica el email del usuario
     */
    @GetMapping("/verify/{token}")
    public void verifyEmail(@PathVariable String token, HttpServletResponse response) throws IOException {
        try {
            userService.verifyEmail(token);
            response.sendRedirect(frontendLoginUrl + "?verified=1");
        } catch (Exception e) {
            log.error("Error en verificación: {}", e.getMessage());
            response.sendRedirect(frontendRegisterUrl + "?verification=failed");
        }
    }

    /**
     * POST /api/auth/refresh - Refresca el JWT token
     */
    @PostMapping("/refresh")
    public ResponseEntity<?> refreshToken(
            Authentication authentication,
            HttpServletResponse response
    ) {
        User user = authenticationHelper.requireAuthenticatedUser(authentication);

        String newToken = jwtProvider.generateToken(user.getEmail());
        addAuthCookie(response, newToken);

        authenticationHelper.logAuthEvent(user.getEmail(), "token_refreshed");

        return ResponseEntity.ok(Map.of(
                "success", true,
                "message", ApiConstants.MSG_TOKEN_REFRESHED,
                "data", Map.of("token", newToken)
        ));
    }

    /**
     * POST /api/auth/logout - Cierra la sesión del usuario
     */
    @PostMapping("/logout")
    public ResponseEntity<?> logout(HttpServletResponse response) {
        removeAuthCookie(response);
        log.info("Logout realizado");

        return ResponseEntity.ok(Map.of(
                "success", true,
                "message", "Logout exitoso"
        ));
    }

    /**
     * Agrega cookie de autenticación segura
     */
    private void addAuthCookie(HttpServletResponse response, String token) {
        // En Spring Boot 3+ (Jakarta), es mejor usar un header Set-Cookie manual 
        // para tener control total sobre SameSite y asegurar que se use el mismo formato.
        String cookieHeader = String.format(
                "%s=%s; Path=%s; Max-Age=%d; HttpOnly; SameSite=%s",
                ApiConstants.JWT_COOKIE_NAME, token, ApiConstants.JWT_COOKIE_PATH,
                ApiConstants.JWT_COOKIE_MAX_AGE, ApiConstants.JWT_SAME_SITE
        );
        
        // Agregar flag Secure si no es localhost o si se desea forzar
        // (En desarrollo local a veces se prefiere false para evitar problemas con HTTP)
        // cookieHeader += "; Secure"; 
        
        response.addHeader("Set-Cookie", cookieHeader);
    }

    /**
     * Elimina la cookie de autenticación
     */
    private void removeAuthCookie(HttpServletResponse response) {
        String cookieHeader = String.format(
                "%s=; Path=%s; Max-Age=0; HttpOnly; SameSite=%s",
                ApiConstants.JWT_COOKIE_NAME, ApiConstants.JWT_COOKIE_PATH,
                ApiConstants.JWT_SAME_SITE
        );
        response.addHeader("Set-Cookie", cookieHeader);
    }
}
